File: tests/test_secrets_store.py
"""敏感信息加密存储测试"""
import json

import pytest

from secrets_store import SecretsStore


@pytest.fixture
def store(tmp_path):
    return SecretsStore(tmp_path / "secrets.json", machine_id_provider=lambda: "test-machine-id")


def test_round_trip(store):
    store.save({"cookie": "test-cookie-value", "extra": 123})
    assert store.get("cookie") == "test-cookie-value"
    assert store.get("extra") == 123


def test_encrypted_on_disk(store):
    store.save({"cookie": "secret"})
    raw = json.loads(store.path.read_text(encoding="utf-8"))
    assert raw.get("_enc") == "fernet-v1"
    assert "data" in raw
    # 密文中不应出现明文 cookie
    assert "secret" not in json.dumps(raw)


def test_wrong_machine_fails(store, tmp_path):
    store.save({"cookie": "secret"})
    evil = SecretsStore(store.path, machine_id_provider=lambda: "different-machine")
    assert evil.get("cookie") is None


def test_plaintext_migration(store):
    """旧版明文 secrets.json 可读取，保存后转为加密"""
    store.path.parent.mkdir(parents=True, exist_ok=True)
    store.path.write_text(json.dumps({"cookie": "plain-cookie"}, ensure_ascii=False), encoding="utf-8")
    assert store.get("cookie") == "plain-cookie"
    store.set("cookie", "plain-cookie")
    raw = json.loads(store.path.read_text(encoding="utf-8"))
    assert raw.get("_enc") == "fernet-v1"

File: tests/licensing/test_secrets_store.cpp
#include <gtest/gtest.h>

#include <QDir>
#include <QTemporaryDir>

#include "secrets_store.h"

using namespace bili;

class SecretsStoreTest : public ::testing::Test {
protected:
    void SetUp() override
    {
        m_dir = std::make_unique<QTemporaryDir>();
        m_path = QDir(m_dir->path()).filePath(QStringLiteral("secrets.json"));
    }

    QString m_path;
    std::unique_ptr<QTemporaryDir> m_dir;
};

TEST_F(SecretsStoreTest, RoundTrip)
{
    SecretsStore store(m_path, []() { return QStringLiteral("test-machine-id"); });
    QJsonObject data;
    data[QStringLiteral("cookie")] = QStringLiteral("test-cookie-value");
    data[QStringLiteral("extra")] = 123;
    store.save(data);

    EXPECT_EQ(store.get(QStringLiteral("cookie")).toString(), QStringLiteral("test-cookie-value"));
    EXPECT_EQ(store.get(QStringLiteral("extra")).toInt(), 123);
}

TEST_F(SecretsStoreTest, EncryptedOnDisk)
{
    SecretsStore store(m_path, []() { return QStringLiteral("test-machine-id"); });
    QJsonObject data;
    data[QStringLiteral("cookie")] = QStringLiteral("secret");
    store.save(data);

    QFile file(m_path);
    ASSERT_TRUE(file.open(QIODevice::ReadOnly | QIODevice::Text));
    const QByteArray raw = file.readAll();
    file.close();

    EXPECT_TRUE(raw.contains("_enc"));
    EXPECT_TRUE(raw.contains("aes-256-gcm-v1"));
    EXPECT_FALSE(raw.contains("secret"));
}

TEST_F(SecretsStoreTest, WrongMachineFails)
{
    {
        SecretsStore store(m_path, []() { return QStringLiteral("test-machine-id"); });
        QJsonObject data;
        data[QStringLiteral("cookie")] = QStringLiteral("secret");
        store.save(data);
    }

    SecretsStore evil(m_path, []() { return QStringLiteral("different-machine"); });
    EXPECT_TRUE(evil.get(QStringLiteral("cookie")).toString().isEmpty());
}

TEST_F(SecretsStoreTest, PlaintextMigration)
{
    QFile file(m_path);
    ASSERT_TRUE(file.open(QIODevice::WriteOnly | QIODevice::Text));
    file.write(R"({"cookie":"plain-cookie"})");
    file.close();

    SecretsStore store(m_path, []() { return QStringLiteral("test-machine-id"); });
    EXPECT_EQ(store.get(QStringLiteral("cookie")).toString(), QStringLiteral("plain-cookie"));

    store.set(QStringLiteral("cookie"), QStringLiteral("plain-cookie"));

    QFile file2(m_path);
    ASSERT_TRUE(file2.open(QIODevice::ReadOnly | QIODevice::Text));
    const QByteArray raw = file2.readAll();
    file2.close();

    EXPECT_TRUE(raw.contains("aes-256-gcm-v1"));
}
